src/main.rs

```diff
@@ -268,3 +275,324 @@ async fn main() {
+            &self.path
+        }
+    }
+
+    impl Drop for TempDir {
+        fn drop(&mut self) {
+            let _ = std::fs::remove_dir_all(&self.path);
+        }
+    }
+
+    const TEST_IMAGE: &str = "ghcr.io/zarf-dev/doom-game:0.0.1";
+    // Split gzip into 1024 * 768 kb chunks
+    const CHUNK_SIZE: usize = 1024 * 768;
+    const ZARF_PAYLOAD_PREFIX: &str = "zarf-payload";
+    // Based on upstream rust-oci-client regex:
+    // https://github.com/oras-project/rust-oci-client/blob/657c1caf9e99ce2184a96aa319fde4f4a8c09439/src/regexp.rs#L3-L5
+    const REFERENCE_REGEXP: &str = r"^((?:(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?::[0-9]+)?/)?[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?::([\w][\w.-]{0,127}))?(?:@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][[:xdigit:]]{32,}))?$";
+
+    #[tokio::test]
+    async fn test_integration() {
+        let docker = Docker::connect_with_socket_defaults()
+            .expect("should have been able to create a Docker client");
+
+        // Create a temporary directory that will auto-cleanup on drop
+        let tmpdir = TempDir::new().expect("should have created temporary directory");
+
+        let env = TestEnv::new(docker.clone(), TEST_IMAGE, tmpdir.path().clone())
+            .await
+            .expect("should have setup the test environment");
+
+        println!("tmpdir is at {:?}", tmpdir.path());
+        let output_root = env.output_dir();
+        unsafe {
+            std::env::set_var("ZARF_INJECTOR_INIT_ROOT", env.input_dir());
+            std::env::set_var("ZARF_INJECTOR_SEED_ROOT", &output_root);
+        }
+        unpack(&env.shasum());
+
+        // Assert the files and directory we expect to exist do exist
+        assert!(Path::new(&output_root.join("index.json")).exists());
+        assert!(Path::new(&output_root.join("manifest.json")).exists());
+        assert!(Path::new(&output_root.join("oci-layout")).exists());
+        assert!(Path::new(&output_root.join("repositories")).exists());
+
+        localize_test_image(TEST_IMAGE, &output_root)
+            .expect("should have localized the test image's index.json");
+
+        // Use :0 to let the operating system decide the random port to listen on
+        let listener = tokio::net::TcpListener::bind("127.0.0.1:0")
+            .await
+            .expect("should have been able to bind listener to a random port on localhost");
+        let random_port = listener
+            .local_addr()
+            .expect("should have been able to resolve the address")
+            .port();
+
+        // Start registry in the background
+        tokio::spawn(async {
+            let app = start_seed_registry();
+            axum::serve(listener, app)
+                .await
+                .expect("should have been able to start serving the registry");
+        });
+
```

Comment:
Do we need a registry readiness check before proceeding?

Reply:
I have not run into any flakes in this regard, but there is no harm to adding a check.